reflex-dom-core/src/Reflex/Dom/Attributes/Types.hs

```diff
@@ -0,0 +1,298 @@
+import Data.Patch
+import Data.Proxy
+import Data.String (IsString(..))
+import Data.Text
+import GHCJS.DOM.Types (JSM)
+import Type.Reflection
+import qualified Data.Dependent.Map as DMap
+import qualified Data.Dependent.Map.Lens as DMap
+import qualified Data.Map as Map
+import qualified GHCJS.DOM.Element as Element
+import qualified GHCJS.DOM.Types as DOM
+
+type Namespace = Text
+
+-- | Attribute names which may or may not have namespaces
+data AttributeName = AttributeName !(Maybe Namespace) !Text deriving (Show, Read, Eq, Ord)
+
+-- | By default, AttributeNames are unnamespaced
+instance IsString AttributeName where
+  fromString = AttributeName Nothing . fromString
+
+-- | Used for controlling 'applyAttrPatchDOM'.
+data PatchSettings
+  = PatchSettings_CheckBeforePatching
+  -- ^ Be careful to minimise the patch. This is used for checking attrs during
+  -- hydration, to prevent touching the DOM when unnecessary. In particular,
+  -- always patching during hydration can result in CSS animations running more
+  -- than once.
+  | PatchSettings_PatchAlways
+  -- ^ Always run the patch. Alternatively: checking before patching is
+  -- optional, and can be omitted to avoid an extra roundtrip.
+
+-- | Define new attributes by creating instances of this class.
+-- This is defined around 'Patch' so we can use 'holdIncremental'.
+-- The 'PatchTarget' is used in declarative attributes (i.e. static and
+-- 'Dynamic' based attributes), and the patch itself is used in modify
+-- attributes ('Event' based attributes).
+class (Monoid (PatchTarget a), Semigroup a, Patch a) => IsAttribute a where
+  -- | Apply an attribute patch to the given DOM element.
+  applyAttrPatchDOM :: PatchSettings -> DOM.Element -> a -> JSM ()
+  -- | Produce a map representing the attributes as text.
+  staticAttrMap :: Proxy a -> PatchTarget a -> Map AttributeName Text -- TODO: is there a reason why we can't make PatchTarget injective? I'd like to get rid of 'Proxy'.
+  -- | Produce a patch by diffing two 'PatchTarget' values. The primary use of
+  -- this is to produce patches for dynamic attributes.
+  diffAttr :: PatchTarget a -> PatchTarget a -> Maybe a
+
+-- | Convert attributes to an attribute patch (by setting all attributes).
+toPatch :: IsAttribute a => PatchTarget a -> Maybe a
+toPatch = diffAttr mempty
+
+-- | The purpose of this is to ensure the ordering of traversal is predictable.
+-- We want to apply the basic attributes before the others such that the basic
+-- ones don't overwrite the finer ones.
+data AttrKey a where
+  AttrKey_BasicAttributes :: AttrKey ModBasicAttributes
+  AttrKey_Other :: IsAttribute a => TypeRep a -> AttrKey a
+
+-- | Does the same job as ArgDict but without actually needing to use ArgDict
+withAttr :: AttrKey a -> (IsAttribute a => b) -> b
+withAttr = \case
+  AttrKey_BasicAttributes -> id
+  AttrKey_Other _ -> id
+
+-- | This is just to apply the 'PatchTarget' type family in DMap values
+newtype PT a = PT { _unPT :: PatchTarget a }
```

Comment:
Does this thing's `Semigroup` instance actually get used?  I think `DMap` is left-biased just like `Map`

Reply:
Explicitly in the `Semigroup` instances of `DeclareAttrs`/`ModifyAttrs`